README.md
LIB-Cairo
=========

Cairo is a 2D graphics library with support for multiple output devices. This is the OS/2 and eComStation port. 

LICENSE
========
GNU Lesser General Public License (LGPL) version 2.1 or the Mozilla Public License (MPL) version 1.1

COMPILE TOOLS
==============
- 

AUTHORS
=============
- Doodle

LINKS
=============
- http://www.cairographics.org/

changelog
2008.05.21.:
 - Updated Cairo to v1.6.4
 - Updated libPixman to 0.10.0
 - Added new OS/2 specific functions:
     cairo_os2_surface_set_draw_offset()
     cairo_os2_surface_get_draw_offset()
     cairo_os2_surface_set_hps()
 - Updated forwarder DLLs to forward to new Cairo DLL
 - Updated FontConfig-emulation code to latest (from Mozilla)

2008.01.18.:
 - Updated Cairo to v1.4.14
 - Added forwarder DLLs so apps linked to old Cairo DLLs can use new DLL

2007.09.19.:
 - Added new OS/2-specific functions to v1.4.6, named it v1.4.7

2007.05.11:
 - Updated Cairo to v1.4.6

2006.09.26:
 - Updated Cairo to v1.2.4

2006.08.09:
 - Updated Cairo to v1.2.2
 - Added FcPatternDel() as a stub to FontConfig/2

2006.08.02:
 - API name changes, requested by original Cairo author

2006.07.30:
 - Replaced the FontConfig+Expat combination with our own small code.
   This should fix fonts.cache problems.

2006.03.14:
 - Modified FontConfig so it won't put fonts.cache* files into your home and
   all of its subfolders, only into PSFONTS folder.
 - Changed compiler to Innotek GCC. This fixes some crashes with DSSaver
   modules, however this makes the DLL dependent on LIBC 06 GA release.

2005.12.22:
 - Implemented dirty_area support. The following Cairo functions will work now:
   cairo_surface_mark_dirty(), cairo_surface_mark_dirty_rectangle()
   You can use them to initialize a newly created surface with pixels from the 
   window (HPS) itself.
 - Updated comments for the APIs in cairo-os2.h file, they were out of date.
   
2005.11.01:
 - Last change was based on old makefile which broke DLL. Fixed makefile.
 - INITINSTANCE and TERMINSTANCE was missing from lnk file, so the DLL did
   not do a cleanup, resulting in huge memory leakage. Fixed.
 - Added new API: cairo_os2_surface_set_HWND()
   See cairo-os2.h for details!
 - Fixed memory leaks in FontConfig.

2005.10.20:
 - Updated code to Cairo 1.0.2 level
 - Modified makefile so it now creates DLL file with version in name
   (e.g.: cairo102.dll)
 - Added two new OS/2-specific API: 
   cairo_os2_initialize() and cairo_os2_uninitialize()
   These are called automatically it Cairo is compiled to be a DLL, but you
   have to call them once if you link to Cairo statically.
 - Separated these two functions into new file to ease the static compilation.
 - Removed libSVG code from DLL as it doesn't really belong to Cairo, and as
   its development has been stopped.

2005.09.03:
 - Created changelog
 - Slightly modified OS2_Patch.c file for the patches to look nicer.
 - Modified makefile, changed compiler switches.
   The adding of '-of' was required in order to make FreeType2 library
   work properly. Without stack frames, the library messes with the
   stack, and the application crashes (mostly when T1 fonts are used).
 - Modified makefile, put a new line to linker file to create more
   verbose map file.


Cairo/config.h
/* config.h for Cairo to be compiled with Netlab's GCC compiler */

/* Make sure we'll use Intel's atomic inc, dec and test and cmpxchg */
/* Update: it's undef'd now, as it does not seem to compile... */
#undef CAIRO_HAS_INTEL_ATOMIC_PRIMITIVES

/* Define to 1 if you have the <dlfcn.h> header file. */
#undef HAVE_DLFCN_H

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the <pthread.h> header file. */
#undef HAVE_PTHREAD_H

/* Define to 1 if you have the `sincos' function. */
#define HAVE_SINCOS

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
/* Modification: Just simply define, and don't define it to be 1! */
/* This is needed so libJPEG can be compiled with this config.h too! */
#define HAVE_STDLIB_H

/* Define to 1 if you have the <strings.h> header file. */
#undef HAVE_STRINGS_H

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the <sys/int_types.h> header file. */
#undef HAVE_SYS_INT_TYPES_H

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if the system has the type `uint128_t'. */
#undef HAVE_UINT128_T

/* Define to 1 if the system has the type `uint64_t'. */
#define HAVE_UINT64_T 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to 1 if you have the `vasnprintf' function. */
#undef HAVE_VASNPRINTF

/* Name of package */
#define PACKAGE "Cairo"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT "http://bugs.freedesktop.org/enter_bug.cgi?product=cairo"

/* Define to the full name of this package. */
#undef PACKAGE_NAME

/* Define to the full name and version of this package. */
#undef PACKAGE_STRING

/* Define to the one symbol short name of this package. */
#undef PACKAGE_TARNAME

/* Define to the version of this package. */
//#undef PACKAGE_VERSION

/* Define to 1 if you have the ANSI C header files. */
#undef STDC_HEADERS

/* Version number of package */
#undef VERSION

/* Define to 1 if your processor stores words with the most significant byte
   first (like Motorola and SPARC, unlike Intel and VAX). */
#undef WORDS_BIGENDIAN

/* Define to 1 if the X Window System is missing or not being used. */
#undef X_DISPLAY_MISSING


/* Some defines for FontConfig */

/* FT_Bitmap_Size structure includes y_ppem field */
#define HAVE_FT_BITMAP_SIZE_Y_PPEM 1

/* Define to 1 if you have the `FT_Get_BDF_Property' function. */
#define HAVE_FT_GET_BDF_PROPERTY 1

/* Define to 1 if you have the `FT_Get_Next_Char' function. */
#define HAVE_FT_GET_NEXT_CHAR 1

/* Define to 1 if you have the `FT_Get_PS_Font_Info' function. */
#define HAVE_FT_GET_PS_FONT_INFO 1

/* Define to 1 if you have the `FT_Get_X11_Font_Format' function. */
#define HAVE_FT_GET_X11_FONT_FORMAT 1

/* Define to 1 if you have the `FT_Has_PS_Glyph_Names' function. */
#define HAVE_FT_HAS_PS_GLYPH_NAMES 1

/* Define to 1 if you have the `memset' function. */
#define HAVE_MEMSET 1


Cairo/OS2_patch.c

/* This file is to patch the Cairo source code so it will have the OS/2 specific changes */

/* Author: Doodle <doodle@scenergy.dfmk.hu> */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#define ADD_CAIRO_CALL_TO_FUNCTIONS

int cairoint(char *path)
{
  FILE *hFile, *hTarget;
  char achBuffer[1024], achBuffer2[1024];
  int bMutexPatchDone;
  int bLineReplaced;
  int iPatched;

  sprintf(achBuffer, "%s\\src\\cairoint.h", path);
  hFile = fopen(achBuffer, "rt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: File [%s] could not be opened for patching!\n", achBuffer);
    return 0;
  }

  sprintf(achBuffer, "%s\\src\\$$$_cairoint.h", path);
  hTarget = fopen(achBuffer, "wt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: Tempfile for [%s] could not be opened for patching!\n", achBuffer);
    fclose(hFile);
    return 0;
  }

  bMutexPatchDone = 0;
  iPatched = 0;
  while (!feof(hFile))
  {
    bLineReplaced = 0;
    achBuffer[0] = 0;
    fgets(achBuffer, sizeof(achBuffer), hFile);
    if (achBuffer[0]==0) break;

    if (strstr(achBuffer, "#if defined(__OS2__)") == achBuffer)
      bMutexPatchDone = 1;

    if ((strstr(achBuffer, "#ifndef CAIRO_MUTEX_DECLARE") == achBuffer) &&
	(!bMutexPatchDone))
    {
      // Include the OS/2 mutex defines!
      fprintf(hTarget, "#if defined(__OS2__) && !defined(CAIRO_MUTEX_DECLARE)\n");
      fprintf(hTarget, "\n");
      fprintf(hTarget, "#define INCL_BASE\n");
      fprintf(hTarget, "#define INCL_PM\n");
      fprintf(hTarget, "#include <os2.h>\n");
      fprintf(hTarget, "\n");
      fprintf(hTarget, "# define CAIRO_MUTEX_DECLARE(name) extern HMTX name\n");
      fprintf(hTarget, "# define CAIRO_MUTEX_DECLARE_GLOBAL(name) extern HMTX name\n");
      fprintf(hTarget, "# define CAIRO_MUTEX_LOCK(name) DosRequestMutexSem(name, SEM_INDEFINITE_WAIT)\n");
      fprintf(hTarget, "# define CAIRO_MUTEX_UNLOCK(name) DosReleaseMutexSem(name)\n");
      fprintf(hTarget, "#endif\n");

      bMutexPatchDone = 1;
      iPatched = 1;
    }

    if (!bLineReplaced)
      fputs(achBuffer, hTarget);
  }

  fclose(hFile);
  fclose(hTarget);

  if (!iPatched)
  {
    // Remove temp file, we did not patch the original one
    sprintf(achBuffer, "%s\\src\\$$$_cairoint.h", path);
    remove(achBuffer);
  } else
  {
    // Remove the original file, then rename the temp file!
    fprintf(stdout, "INFO: File [%s\\src\\cairoint.h] has been patched!\n", path);
    sprintf(achBuffer, "%s\\src\\cairoint.h", path);
    sprintf(achBuffer2, "%s\\src\\$$$_cairoint.h", path);
    remove(achBuffer);
    rename(achBuffer2, achBuffer);
  }
  return 1;
}

int cairoh(char *path)
{
  FILE *hFile, *hTarget;
  char achBuffer[1024], achBuffer2[1024];
  int iPatched, bSurfacePatchDone, bLineReplaced;

  sprintf(achBuffer, "%s\\src\\cairo.h", path);
  hFile = fopen(achBuffer, "rt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: File [%s] could not be opened for patching!\n", achBuffer);
    return 0;
  }

  sprintf(achBuffer, "%s\\src\\$$$_cairo.h", path);
  hTarget = fopen(achBuffer, "wt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: Tempfile for [%s] could not be opened for patching!\n", achBuffer);
    fclose(hFile);
    return 0;
  }

  iPatched = 0;
  bSurfacePatchDone = 0;
  while (!feof(hFile))
  {
    bLineReplaced = 0;
    achBuffer[0] = 0;
    fgets(achBuffer, sizeof(achBuffer), hFile);
    if (achBuffer[0]==0) break;

    if (strstr(achBuffer, "CAIRO_SURFACE_TYPE_OS2"))
      bSurfacePatchDone = 1;

    if ((strstr(achBuffer, "} cairo_surface_type_t;")) &&
	(!bSurfacePatchDone))
    {
      // Include the OS/2 surface type!
      fprintf(hTarget, "    , CAIRO_SURFACE_TYPE_OS2\n");
      bSurfacePatchDone = 1;
      iPatched = 1;
    }

    if (!bLineReplaced)
      fputs(achBuffer, hTarget);
  }

  fclose(hFile);
  fclose(hTarget);

  if (!iPatched)
  {
    // Remove temp file, we did not patch the original one
    sprintf(achBuffer, "%s\\src\\$$$_cairo.h", path);
    remove(achBuffer);
  } else
  {
    // Remove the original file, then rename the temp file!
    fprintf(stdout, "INFO: File [%s\\src\\cairo.h] has been patched!\n", path);
    sprintf(achBuffer, "%s\\src\\cairo.h", path);
    sprintf(achBuffer2, "%s\\src\\$$$_cairo.h", path);
    remove(achBuffer);
    rename(achBuffer2, achBuffer);
  }
  return 1;
}

int icimage(char *path)
{
  FILE *hFile, *hTarget;
  char achBuffer[1024], achBuffer2[1024];
  int bLineReplaced;
  int iPatched;

  sprintf(achBuffer, "%s\\pixman\\src\\icimage.c", path);
  hFile = fopen(achBuffer, "rt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: File [%s] could not be opened for patching!\n", achBuffer);
    return 0;
  }

  sprintf(achBuffer, "%s\\pixman\\src\\$$$_icimage.c", path);
  hTarget = fopen(achBuffer, "wt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: Tempfile for [%s] could not be opened for patching!\n", achBuffer);
    fclose(hFile);
    return 0;
  }

  iPatched = 0;
  while (!feof(hFile))
  {
    bLineReplaced = 0;
    achBuffer[0] = 0;
    fgets(achBuffer, sizeof(achBuffer), hFile);
    if (achBuffer[0]==0) break;

    if (strstr(achBuffer, "uint ") != NULL)
    {
      char *pchPos;

      /* This line has to be pathed! */

      while ((pchPos = strstr(achBuffer, "uint ")) != NULL)
      {
        char *pchTemp;
        pchPos += 4;

        pchTemp = pchPos;
        while (*pchTemp) pchTemp++;
        pchTemp+=4;

        while (pchTemp>pchPos)
        {
          pchTemp[0] = pchTemp[-4];
          pchTemp--;
        }
        pchPos[0] = '3';
        pchPos[1] = '2';
        pchPos[2] = '_';
        pchPos[3] = 't';
      }

      fputs(achBuffer, hTarget);
      bLineReplaced = 1;
      iPatched = 1;
    }

    if (!bLineReplaced)
      fputs(achBuffer, hTarget);
  }

  fclose(hFile);
  fclose(hTarget);

  if (!iPatched)
  {
    // Remove temp file, we did not patch the original one
    sprintf(achBuffer, "%s\\pixman\\src\\$$$_icimage.c", path);
    remove(achBuffer);
  } else
  {
    // Remove the original file, then rename the temp file!
    fprintf(stdout, "INFO: File [%s\\pixman\\src\\icimage.c] has been patched!\n", path);
    sprintf(achBuffer, "%s\\pixman\\src\\icimage.c", path);
    sprintf(achBuffer2, "%s\\pixman\\src\\$$$_icimage.c", path);
    remove(achBuffer);
    rename(achBuffer2, achBuffer);
  }
  return 1;
}

int cairo_funcprep(char *path, char *file, char *funcbegin)
{
#ifdef ADD_CAIRO_CALL_TO_FUNCTIONS
  FILE *hFile, *hTarget;
  char achBuffer[1024], achBuffer2[1024];
  int bLineReplaced;
  int iPatched;
  int bFirstLine;

  sprintf(achBuffer, "%s\\src\\%s", path, file);
  hFile = fopen(achBuffer, "rt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: File [%s] could not be opened for patching!\n", achBuffer);
    return 0;
  }

  sprintf(achBuffer, "%s\\src\\$$$_%s", path, file);
  hTarget = fopen(achBuffer, "wt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: Tempfile for [%s] could not be opened for patching!\n", achBuffer);
    fclose(hFile);
    return 0;
  }

  bFirstLine = 1;
  iPatched = 0;
  bLineReplaced = 0;
  sprintf(achBuffer, "");
  while (!feof(hFile))
  {
    // Copy previous line into buffer2
    strcpy(achBuffer2, achBuffer);

    // Read line
    achBuffer[0] = 0;
    fgets(achBuffer, sizeof(achBuffer), hFile);
    if (achBuffer[0]==0) break;

    if ((strstr(achBuffer, funcbegin) == achBuffer) &&
        (strstr(achBuffer, "cairo_public ")!=achBuffer) &&
	(strstr(achBuffer, "(")))
    {
      /* This is a cairo function define! */
      /* Let's see if we have to patch it or not! */

      if ((strstr(achBuffer2, "cairo_public") != achBuffer2) ||
          (strstr(achBuffer, "cairo_call") != achBuffer))
      {
        /* Either the first or second line has to be patched! */

        if (strstr(achBuffer2, "cairo_public") == achBuffer2)
        {
          /* It already has cairo_public, so don't put it there again */
          fprintf(hTarget, "%s", achBuffer2);
        } else
        {
          fprintf(hTarget, "cairo_public %s", achBuffer2);
        }

        if (strstr(achBuffer, "cairo_call") == achBuffer)
        {
          /* It already has cairo_call, so don't put it there again */
          fprintf(hTarget, "%s", achBuffer);
        } else
        {
          fprintf(hTarget, "cairo_call   %s", achBuffer);
        }

        bLineReplaced += 2;
        iPatched = 1;
      }
    }

    if ((strstr(achBuffer, "typedef ") == achBuffer) &&
	(strstr(achBuffer, "(*")) &&
	(strstr(achBuffer, "cairo_call") == NULL)
       )
    {
      char *pchType, *pchRemaining;

      // This is a cairo callback typedef!
      pchType = &(achBuffer[8]);
      pchRemaining = strstr(achBuffer, "(*");
      *(pchRemaining-1) = 0;

      fputs(achBuffer2, hTarget);
      fprintf(hTarget, "typedef %s cairo_call %s", pchType, pchRemaining);

      bLineReplaced += 2;
      iPatched = 1;
    }

    if ((!bLineReplaced) &&
        (!bFirstLine))
      fputs(achBuffer2, hTarget);

    if (bLineReplaced>0)
      bLineReplaced--;
    bFirstLine = 0;
  }

  if (!bLineReplaced)
    fputs(achBuffer2, hTarget);

  fclose(hFile);
  fclose(hTarget);

  if (!iPatched)
  {
    // Remove temp file, we did not patch the original one
    sprintf(achBuffer, "%s\\src\\$$$_%s", path, file);
    remove(achBuffer);
  } else
  {
    // Remove the original file, then rename the temp file!
    fprintf(stdout, "INFO: File [%s\\src\\%s] has been patched!\n", path, file);
    sprintf(achBuffer, "%s\\src\\%s", path, file);
    sprintf(achBuffer2, "%s\\src\\$$$_%s", path, file);
    remove(achBuffer);
    rename(achBuffer2, achBuffer);
  }
#endif
  return 1;
}

int cairo_funcprep2(char *path, char *file, char *funcname)
{
#ifdef ADD_CAIRO_CALL_TO_FUNCTIONS
  FILE *hFile, *hTarget;
  char achBuffer[1024], achBuffer2[1024];
  int bLineReplaced;
  int iPatched;

  sprintf(achBuffer, "%s\\src\\%s", path, file);
  hFile = fopen(achBuffer, "rt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: File [%s] could not be opened for patching!\n", achBuffer);
    return 0;
  }

  sprintf(achBuffer, "%s\\src\\$$$_%s", path, file);
  hTarget = fopen(achBuffer, "wt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: Tempfile for [%s] could not be opened for patching!\n", achBuffer);
    fclose(hFile);
    return 0;
  }

  iPatched = 0;
  bLineReplaced = 0;
  sprintf(achBuffer, "");
  while (!feof(hFile))
  {
    // Read line
    achBuffer[0] = 0;
    fgets(achBuffer, sizeof(achBuffer), hFile);
    if (achBuffer[0]==0) break;

    if (strstr(achBuffer, funcname) == achBuffer)
    {
      fprintf(hTarget, "cairo_call  %s", achBuffer);

      bLineReplaced += 1;
      iPatched = 1;
    }

    if (!bLineReplaced)
      fputs(achBuffer, hTarget);

    if (bLineReplaced>0)
      bLineReplaced--;
  }

  fclose(hFile);
  fclose(hTarget);

  if (!iPatched)
  {
    // Remove temp file, we did not patch the original one
    sprintf(achBuffer, "%s\\src\\$$$_%s", path, file);
    remove(achBuffer);
  } else
  {
    // Remove the original file, then rename the temp file!
    fprintf(stdout, "INFO: File [%s\\src\\%s] has been patched!\n", path, file);
    sprintf(achBuffer, "%s\\src\\%s", path, file);
    sprintf(achBuffer2, "%s\\src\\$$$_%s", path, file);
    remove(achBuffer);
    rename(achBuffer2, achBuffer);
  }
  return 1;
}

int ftconfig(char *path)
{
  FILE *hFile, *hTarget;
  char achBuffer[1024], achBuffer2[1024];
  int iPatched;

  sprintf(achBuffer, "%s\\include\\freetype\\config\\ftconfig.h", path);
  hFile = fopen(achBuffer, "rt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: File [%s] could not be opened for patching!\n", achBuffer);
    return 0;
  }

  sprintf(achBuffer, "%s\\include\\freetype\\config\\$$$_ftconfig.h", path);
  hTarget = fopen(achBuffer, "wt");
  if (!hFile)
  {
    fprintf(stdout, "INFO: Tempfile for [%s] could not be opened for patching!\n", achBuffer);
    fclose(hFile);
    return 0;
  }

  iPatched = 0;
  while (!feof(hFile))
  {
    achBuffer[0] = 0;
    fgets(achBuffer, sizeof(achBuffer), hFile);
    if (achBuffer[0]==0) break;

    if (strstr(achBuffer, "Watcom doesn't provide 64-bit data types"))
    {
      // Found it! Replace it!
      fprintf(hTarget, "  /* this compiler provides the __int64 type */\n");
      fprintf(hTarget, "#define FT_LONG64\n");
      fprintf(hTarget, "#define FT_INT64 __int64\n");

      iPatched = 1;
    }
    else
      fputs(achBuffer, hTarget);
  }

  fclose(hFile);
  fclose(hTarget);

  if (!iPatched)
  {
    // Remove temp file, we did not patch the original one
    sprintf(achBuffer, "%s\\include\\freetype\\config\\$$$_ftconfig.h", path);
    remove(achBuffer);
  } else
  {
    // Remove the original file, then rename the temp file!
    fprintf(stdout, "INFO: File [%s\\include\\freetype\\config\\ftconfig.h] has been patched!\n", path);
    sprintf(achBuffer, "%s\\include\\freetype\\config\\ftconfig.h", path);
    sprintf(achBuffer2, "%s\\include\\freetype\\config\\$$$_ftconfig.h", path);
    remove(achBuffer);
    rename(achBuffer2, achBuffer);
  }
#endif
  return 1;
}

int main(int argc, char *argv[])
{
  if (argc<=2)
  {
    printf("Cairo Source Code Patcher (for cairo-1.2.0)\n");
    printf("Please give me the base directory of\n");
    printf("Cairo and Freetype2 sources\n");
    printf("as two command line parameters!\n");
    return 1;
  }

  // Patch cairoint.h
  if (!cairoint(argv[1]))
    return 1;

  // Patch cairo.h to extend cairo_surface_t with CAIRO_SURFACE_TYPE_OS2
  if (!cairoh(argv[1]))
    return 1;

  // Patch cairo.h
  if (!cairo_funcprep(argv[1], "cairo.h", "cairo_"))
    return 1;

  // Patch cairo-ft.h
  if (!cairo_funcprep(argv[1], "cairo-ft.h", "cairo_"))
    return 1;

  // Patch cairo-ps.h
  if (!cairo_funcprep(argv[1], "cairo-ps.h", "cairo_"))
    return 1;

  // Patch cairo-pdf.h
  if (!cairo_funcprep(argv[1], "cairo-pdf.h", "cairo_"))
    return 1;


  // Patch cairo-svg.h
  if (!cairo_funcprep(argv[1], "cairo-svg.h", "cairo_"))
    return 1;

  // Patch cairo-output-stream.c
  if (!cairo_funcprep2(argv[1], "cairo-output-stream.c", "stdio_write"))
    return 1;

  // Patch cairo-os2.h
  if (!cairo_funcprep(argv[1], "cairo-os2.h", "cairo_"))
    return 1;

  // Patch cairo-os2-surface.c
  if (!cairo_funcprep(argv[1], "cairo-os2-surface.c", "cairo_"))
    return 1;

  /*
  // Patch cairo-scaled-font.c
  if (!cairo_funcprep2(argv[1], "cairo-scaled-font.c", "_cairo_scaled_glyph_destroy"))
    return 1;

  // Patch cairo-svg-surface.c
  if (!cairo_funcprep2(argv[1], "cairo-svg-surface.c", "base64_write_func"))
    return 1;
  */
  /*
  // Patch icimage.c
  if (!icimage(argv[1]))
    return 1;

  // Patch ftconfig.h
  if (!ftconfig(argv[2]))
    return 1;
  */
  return 0;
}



Cairo/cairo-1.6.4/src/cairo-os2-private.h
/* vim: set sw=4 sts=4 et cin: */
/* cairo - a vector graphics library with display and print output
 *
 * Copyright (c) 2005-2006 netlabs.org
 *
 * This library is free software; you can redistribute it and/or
 * modify it either under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation
 * (the "LGPL") or, at your option, under the terms of the Mozilla
 * Public License Version 1.1 (the "MPL"). If you do not alter this
 * notice, a recipient may use your version of this file under either
 * the MPL or the LGPL.
 *
 * You should have received a copy of the LGPL along with this library
 * in the file COPYING-LGPL-2.1; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 * You should have received a copy of the MPL along with this library
 * in the file COPYING-MPL-1.1
 *
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY
 * OF ANY KIND, either express or implied. See the LGPL or the MPL for
 * the specific language governing rights and limitations.
 *
 * The Original Code is the cairo graphics library.
 *
 * The Initial Developer of the Original Code is
 *     Doodle <doodle@scenergy.dfmk.hu>
 *
 * Contributor(s):
 *     Peter Weilbacher <mozilla@Weilbacher.org>
 */

#ifndef CAIRO_OS2_PRIVATE_H
#define CAIRO_OS2_PRIVATE_H

#define INCL_DOS
#define INCL_DOSSEMAPHORES
#define INCL_DOSERRORS
#define INCL_WIN
#define INCL_GPI
#ifdef __WATCOMC__
# include <os2.h>
#else
# include <os2emx.h>
#endif

#include <cairo-os2.h>
#include "cairoint.h"

typedef struct _cairo_os2_surface
{
    cairo_surface_t        base;

    /* Mutex semaphore to protect private fields from concurrent access */
    HMTX                   hmtx_use_private_fields;
    /* Private fields: */
    HPS                    hps_client_window;
    HWND                   hwnd_client_window;
    BITMAPINFO2            bitmap_info;
    unsigned char         *pixels;
    cairo_image_surface_t *image_surface;
    int                    pixel_array_lend_count;
    HEV                    hev_pixel_array_came_back;

    RECTL                  rcl_dirty_area;
    cairo_bool_t           dirty_area_present;

    /* General flags: */
    cairo_bool_t           blit_as_changes;
    int                    horz_draw_offset;
    int                    vert_draw_offset;
} cairo_os2_surface_t;

#endif /* CAIRO_OS2_PRIVATE_H */


Cairo/cairo-1.6.4/src/cairo-os2.h
/* vim: set sw=4 sts=4 et cin: */
/* cairo - a vector graphics library with display and print output
 *
 * Copyright (c) 2005-2006 netlabs.org
 *
 * This library is free software; you can redistribute it and/or
 * modify it either under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation
 * (the "LGPL") or, at your option, under the terms of the Mozilla
 * Public License Version 1.1 (the "MPL"). If you do not alter this
 * notice, a recipient may use your version of this file under either
 * the MPL or the LGPL.
 *
 * You should have received a copy of the LGPL along with this library
 * in the file COPYING-LGPL-2.1; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 * You should have received a copy of the MPL along with this library
 * in the file COPYING-MPL-1.1
 *
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY
 * OF ANY KIND, either express or implied. See the LGPL or the MPL for
 * the specific language governing rights and limitations.
 *
 * The Original Code is the cairo graphics library.
 *
 * The Initial Developer of the Original Code is
 *     Doodle <doodle@scenergy.dfmk.hu>
 *
 * Contributor(s):
 *     Peter Weilbacher <mozilla@Weilbacher.org>
 */

#ifndef _CAIRO_OS2_H_
#define _CAIRO_OS2_H_

#include <cairo.h>

CAIRO_BEGIN_DECLS

/* The OS/2 Specific Cairo API */

cairo_public void
cairo_os2_init (void);

cairo_public void
cairo_os2_fini (void);

#if CAIRO_HAS_OS2_SURFACE

cairo_public cairo_surface_t *
cairo_os2_surface_create (HPS hps_client_window,
                          int width,
                          int height);

cairo_public void
cairo_os2_surface_set_hwnd (cairo_surface_t *surface,
                            HWND             hwnd_client_window);

cairo_public int
cairo_os2_surface_set_size (cairo_surface_t *surface,
                            int              new_width,
                            int              new_height,
                            int              timeout);

cairo_public void
cairo_os2_surface_refresh_window (cairo_surface_t *surface,
                                  HPS              hps_begin_paint,
                                  PRECTL           prcl_begin_paint_rect);

cairo_public void
cairo_os2_surface_set_manual_window_refresh (cairo_surface_t *surface,
                                             cairo_bool_t     manual_refresh);

cairo_public cairo_bool_t
cairo_os2_surface_get_manual_window_refresh (cairo_surface_t *surface);

cairo_public void
cairo_os2_surface_set_draw_offset (cairo_surface_t *surface,
                                   int horz_offset, int vert_offset);

cairo_public void
cairo_os2_surface_get_draw_offset (cairo_surface_t *surface,
                                   int *horz_offset, int *vert_offset);

cairo_public void
cairo_call   cairo_os2_surface_set_hps (cairo_surface_t *surface,
                                   HPS hps);

#else  /* CAIRO_HAS_OS2_SURFACE */
# error Cairo was not compiled with support for the OS/2 backend
#endif /* CAIRO_HAS_OS2_SURFACE */

CAIRO_END_DECLS

#endif /* _CAIRO_OS2_H_ */
